File: clients/DiscoKit/Sources/DiscoKit/AutoUpload/UploadJournal.swift
import Foundation
import GRDB

/// What happened to an asset the journal knows about.
public enum UploadState: String, Sendable {
    case sent
    /// Was already in the library when auto-upload was switched on.
    case skippedPreexisting = "skipped-preexisting"
    /// Failed; will be retried on a later pass until the attempt cap.
    case deferred
}

public struct JournalEntry: Sendable {
    public let assetID: String
    public let serverName: String
    public let state: UploadState
    public let error: String?
    public let at: Date

    public init(assetID: String, serverName: String, state: UploadState, error: String?, at: Date) {
        self.assetID = assetID; self.serverName = serverName
        self.state = state; self.error = error; self.at = at
    }
}

public struct JournalCounts: Sendable {
    public let sent: Int
    public let skipped: Int
    public let deferred: Int

    public init(sent: Int, skipped: Int, deferred: Int) {
        self.sent = sent; self.skipped = skipped; self.deferred = deferred
    }
}

/// Remembers which photos have already been dealt with, so auto-upload never sends the same
/// one twice.
///
/// This is the source of truth for "was it uploaded" — deliberately NOT a diff against the
/// server. If a photo were re-sent whenever it went missing on the server, deleting it in
/// the web UI would just make the phone put it back.
///
/// Identity is the asset's local identifier plus its modification date: editing a photo in
/// Photos bumps that date, and the edited version is genuinely new content.
public final class UploadJournal: @unchecked Sendable {   // dbQueue (GRDB) is internally synchronized
    private let dbQueue: DatabaseQueue

    public init(dbQueue: DatabaseQueue) throws {
        self.dbQueue = dbQueue
        try migrate()
    }

    public convenience init(path: String) throws {
        try self.init(dbQueue: try DatabaseQueue(path: path))
    }

    private func migrate() throws {
        try dbQueue.write { db in
            try db.execute(sql: """
                CREATE TABLE IF NOT EXISTS uploads(
                  asset_id TEXT PRIMARY KEY,
                  modified INTEGER NOT NULL,
                  bytes INTEGER NOT NULL DEFAULT 0,
                  sha TEXT,
                  server_name TEXT,
                  state TEXT NOT NULL,
                  attempts INTEGER NOT NULL DEFAULT 0,
                  error TEXT,
                  at INTEGER NOT NULL
                );
                CREATE INDEX IF NOT EXISTS idx_uploads_state ON uploads(state);
            """)
        }
    }

    /// True when this exact version of the asset was already handled. A deferred asset is
    /// NOT known: it is meant to be retried.
    public func isKnown(assetID: String, modified: Date) throws -> Bool {
        try dbQueue.read { db in
            let row = try Row.fetchOne(db, sql: """
                SELECT state FROM uploads WHERE asset_id = ? AND modified = ?
            """, arguments: [assetID, Int64(modified.timeIntervalSince1970 * 1000)])
            guard let row else { return false }
            return (row["state"] as String?) != UploadState.deferred.rawValue
        }
    }

    public func attempts(assetID: String) throws -> Int {
        try dbQueue.read { db in
            try Int.fetchOne(db, sql: "SELECT attempts FROM uploads WHERE asset_id = ?",
                             arguments: [assetID]) ?? 0
        }
    }

    /// Records an upload. `modified` must be the value the upload actually read, not
    /// today's: an asset edited mid-upload would otherwise be stored under its NEW identity
    /// and the edit would never be sent.
    public func markSent(assetID: String, modified: Date, bytes: Int64,
                         sha: String?, serverName: String) throws {
        try put(assetID: assetID, modified: modified, bytes: bytes, sha: sha,
                serverName: serverName, state: .sent, error: nil, attempts: 0)
    }

    public func markDeferred(assetID: String, modified: Date, error: String) throws {
        let n = try attempts(assetID: assetID) + 1
        try put(assetID: assetID, modified: modified, bytes: 0, sha: nil,
                serverName: nil, state: .deferred, error: error, attempts: n)
    }

    /// Records what is already in the library without uploading it. This is what makes
    /// "new photos only" hold when the feature is switched on over years of pictures.
    public func seedPreexisting(_ assets: [(id: String, modified: Date)]) throws {
        try dbQueue.write { db in
            for a in assets {
                try db.execute(sql: """
                    INSERT INTO uploads(asset_id, modified, bytes, sha, server_name, state, attempts, error, at)
                    VALUES (?, ?, 0, NULL, NULL, ?, 0, NULL, ?)
                    ON CONFLICT(asset_id) DO UPDATE SET modified = excluded.modified,
                        state = excluded.state, at = excluded.at
                """, arguments: [a.id, Int64(a.modified.timeIntervalSince1970 * 1000),
                                 UploadState.skippedPreexisting.rawValue,
                                 Int64(Date().timeIntervalSince1970 * 1000)])
            }
        }
    }

    /// Forgets the "was already there" marks, turning the existing library back into work.
    /// Uploaded and deferred rows are left alone: what already went up must not go again.
    /// Returns how many photos are now waiting.
    @discardableResult
    public func unseed() throws -> Int {
        try dbQueue.write { db in
            let n = try Int.fetchOne(db, sql: "SELECT COUNT(*) FROM uploads WHERE state = ?",
                                     arguments: [UploadState.skippedPreexisting.rawValue]) ?? 0
            try db.execute(sql: "DELETE FROM uploads WHERE state = ?",
                           arguments: [UploadState.skippedPreexisting.rawValue])
            return n
        }
    }

    public func counts() throws -> JournalCounts {
        try dbQueue.read { db in
            func n(_ state: UploadState) throws -> Int {
                try Int.fetchOne(db, sql: "SELECT COUNT(*) FROM uploads WHERE state = ?",
                                 arguments: [state.rawValue]) ?? 0
            }
            return JournalCounts(sent: try n(.sent), skipped: try n(.skippedPreexisting),
                                 deferred: try n(.deferred))
        }
    }

    public func recent(limit: Int) throws -> [JournalEntry] {
        try dbQueue.read { db in
            try Row.fetchAll(db, sql: """
                SELECT asset_id, server_name, state, error, at FROM uploads
                ORDER BY at DESC LIMIT ?
            """, arguments: [limit]).map { row in
                JournalEntry(
                    assetID: row["asset_id"],
                    serverName: row["server_name"] ?? "",
                    state: UploadState(rawValue: row["state"] ?? "") ?? .deferred,
                    error: row["error"],
                    at: Date(timeIntervalSince1970: Double(row["at"] as Int64) / 1000),
                )
            }
        }
    }

    private func put(assetID: String, modified: Date, bytes: Int64, sha: String?,
                     serverName: String?, state: UploadState, error: String?, attempts: Int) throws {
        try dbQueue.write { db in
            try db.execute(sql: """
                INSERT INTO uploads(asset_id, modified, bytes, sha, server_name, state, attempts, error, at)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
                ON CONFLICT(asset_id) DO UPDATE SET
                    modified = excluded.modified, bytes = excluded.bytes, sha = excluded.sha,
                    server_name = excluded.server_name, state = excluded.state,
                    attempts = excluded.attempts, error = excluded.error, at = excluded.at
            """, arguments: [assetID, Int64(modified.timeIntervalSince1970 * 1000), bytes, sha,
                             serverName, state.rawValue, attempts, error,
                             Int64(Date().timeIntervalSince1970 * 1000)])
        }
    }
}

File: clients/android-discodrive/app/src/main/java/org/discodrive/android/autoupload/UploadJournal.kt
package org.discodrive.android.autoupload

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import java.io.File

/** What happened to a file the journal knows about. */
const val STATE_SENT = "sent"
const val STATE_SKIPPED = "skipped-preexisting"
const val STATE_DEFERRED = "deferred"

data class JournalEntry(
    val path: String,
    val serverName: String,
    val state: String,
    val error: String?,
    val at: Long,
)

data class JournalCounts(val sent: Int, val skipped: Int, val deferred: Int)

/**
 * Remembers which local files have already been dealt with, so auto-upload never sends the
 * same photo twice.
 *
 * This is the source of truth for "was it uploaded" — deliberately NOT a diff against the
 * server. If a file were re-sent whenever it went missing on the server, deleting a photo
 * in the web UI would just make the phone put it back.
 *
 * Identity is (path, size, mtime): a file edited in place gets a new mtime and is treated
 * as new content, which is what the user means by "I changed this photo".
 */
class UploadJournal(context: Context) {

    private val helper = object : SQLiteOpenHelper(context, "autoupload.db", null, 1) {
        override fun onCreate(db: SQLiteDatabase) {
            db.execSQL(
                """
                CREATE TABLE sent(
                  path TEXT PRIMARY KEY,
                  size INTEGER NOT NULL,
                  mtime INTEGER NOT NULL,
                  sha TEXT,
                  server_name TEXT,
                  state TEXT NOT NULL,
                  attempts INTEGER NOT NULL DEFAULT 0,
                  error TEXT,
                  at INTEGER NOT NULL
                )
                """.trimIndent()
            )
            db.execSQL("CREATE INDEX sent_state ON sent(state)")
        }

        override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) = Unit
    }

    /**
     * True when this exact file (same path, size and mtime) was already handled and should
     * not be uploaded again. A deferred file is NOT known: it is meant to be retried.
     */
    fun isKnown(file: File): Boolean = helper.readableDatabase.rawQuery(
        "SELECT state FROM sent WHERE path=? AND size=? AND mtime=?",
        arrayOf(key(file), file.length().toString(), file.lastModified().toString()),
    ).use { c -> c.moveToFirst() && c.getString(0) != STATE_DEFERRED }

    /** How many attempts this file has already cost, for the give-up rule. */
    fun attempts(file: File): Int = helper.readableDatabase.rawQuery(
        "SELECT attempts FROM sent WHERE path=?", arrayOf(key(file)),
    ).use { c -> if (c.moveToFirst()) c.getInt(0) else 0 }

    fun markSent(file: File, serverName: String, sha: String?) =
        put(file, serverName, sha, STATE_SENT, error = null, attempts = 0)

    fun markDeferred(file: File, error: String) =
        put(file, serverName = null, sha = null, state = STATE_DEFERRED, error = error, attempts = attempts(file) + 1)

    /**
     * Records everything currently in the source folder as pre-existing, without uploading
     * it. This is what makes "new files only" hold when a rule is switched on over a folder
     * that already holds an archive.
     */
    fun seedPreexisting(files: List<File>) {
        val db = helper.writableDatabase
        db.beginTransaction()
        try {
            for (f in files) put(f, serverName = null, sha = null, state = STATE_SKIPPED, error = null, attempts = 0, db = db)
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }
    }

    fun counts(): JournalCounts {
        fun n(state: String) = helper.readableDatabase.rawQuery(
            "SELECT COUNT(*) FROM sent WHERE state=?", arrayOf(state),
        ).use { c -> if (c.moveToFirst()) c.getInt(0) else 0 }
        return JournalCounts(sent = n(STATE_SENT), skipped = n(STATE_SKIPPED), deferred = n(STATE_DEFERRED))
    }

    fun recent(limit: Int): List<JournalEntry> = helper.readableDatabase.rawQuery(
        "SELECT path, server_name, state, error, at FROM sent ORDER BY at DESC LIMIT ?",
        arrayOf(limit.toString()),
    ).use { c ->
        buildList {
            while (c.moveToNext()) {
                add(JournalEntry(c.getString(0), c.getString(1) ?: "", c.getString(2), c.getString(3), c.getLong(4)))
            }
        }
    }

    fun close() = helper.close()

    /** Journal key: the same folder reachable as /sdcard/... and /storage/emulated/0/...
     *  must not produce two entries for one file. */
    private fun key(file: File): String = Rule.normalize(file.path)

    private fun put(
        file: File,
        serverName: String?,
        sha: String?,
        state: String,
        error: String?,
        attempts: Int,
        db: SQLiteDatabase = helper.writableDatabase,
    ) {
        val v = ContentValues().apply {
            put("path", key(file))
            put("size", file.length())
            put("mtime", file.lastModified())
            put("sha", sha)
            put("server_name", serverName)
            put("state", state)
            put("attempts", attempts)
            put("error", error)
            put("at", System.currentTimeMillis())
        }
        db.insertWithOnConflict("sent", null, v, SQLiteDatabase.CONFLICT_REPLACE)
    }
}
